File: app/src/main/java/tech/pauly/findapet/shared/events/PermissionEvent.java
package tech.pauly.findapet.shared.events;

import java.util.Arrays;
import java.util.Objects;

import tech.pauly.findapet.shared.PermissionHelper;

public class PermissionEvent extends BaseViewEvent {

    private String[] permissions;
    private PermissionListener listener;
    private int requestCode;

    public static PermissionEvent build(Object emitter) {
        return new PermissionEvent(emitter.getClass());
    }

    private PermissionEvent(Class emitter) {
        this.emitter = emitter;
    }

    public PermissionEvent requestPermission(String... permissions) {
        this.permissions = permissions;
        return this;
    }

    public PermissionEvent listener(PermissionListener listener) {
        this.listener = listener;
        return this;
    }

    public PermissionEvent code(int requestCode) {
        this.requestCode = requestCode;
        return this;
    }

    public String[] getPermissions() {
        return permissions;
    }
    public PermissionListener getListener() {
        return listener;
    }

    public int getRequestCode() {
        return requestCode;
    }

    public interface PermissionListener {
        void onPermissionResult(PermissionHelper.PermissionRequestResponse response);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        PermissionEvent that = (PermissionEvent) o;
        return requestCode == that.requestCode && Arrays.equals(permissions, that.permissions) && Objects.equals(listener, that.listener);
    }

    @Override
    public int hashCode() {

        int result = Objects.hash(listener, requestCode);
        result = 31 * result + Arrays.hashCode(permissions);
        return result;
    }
}

File: app/src/main/java/tech/pauly/findapet/shared/events/PermissionEvent.kt
package tech.pauly.findapet.shared.events

import tech.pauly.findapet.shared.PermissionRequestResponse
import java.util.*

open class PermissionEvent(private val emitter: Any,
                           open val permissions: Array<String>,
                           open val listener: PermissionListener,
                           open val requestCode: Int) : BaseViewEvent(emitter.javaClass) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as PermissionEvent

        if (emitter != other.emitter) return false
        if (!Arrays.equals(permissions, other.permissions)) return false
        if (listener != other.listener) return false
        if (requestCode != other.requestCode) return false

        return true
    }

    override fun hashCode(): Int {
        var result = emitter.hashCode()
        result = 31 * result + Arrays.hashCode(permissions)
        result = 31 * result + listener.hashCode()
        result = 31 * result + requestCode
        return result
    }
}

interface PermissionListener {
    fun onPermissionResult(response: PermissionRequestResponse)
}